### PR TITLE
APP-156: [C++] Basic level Apps: Data collection

### PR DESCRIPTION
**Pull request by RostakaGmfun**
/cc @Dubland @Slash32 @vadimol

### CODE REVIEW

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp [hunk 1]
```diff
@@ -14,77 +14,110 @@
...
-     */
-    auto kaaClient =  Kaa::newClient();
+class TemperatureSensor : public IConfigurationReceiver {
+public:
+    TemperatureSensor():
+        kaaClient_(Kaa::newClient()),
+        interval_(samplePeriod_ = 0),
+        timer_(service_, interval_)
+    {
+        // Set a custom strategy for uploading logs.
```
Comment (Slash32) [inline datacollectiondemo/source/cpp/src/KaaDemo.cpp:40]: Could an information for why do we need this be added to the comment as well?

**Comment by Slash32** [comment]
Minor comment, otherwise LGTM

File: datacollectiondemo/source/cpp/build.sh [hunk 2]
```diff
@@ -58,50 +73,19 @@ function select_arch {
...
-}
-
 function build_app {
     cd $PROJECT_HOME &&
     mkdir -p "$PROJECT_HOME/$BUILD_DIR" &&
-    cp "$KAA_LIB_PATH/$BUILD_DIR/"libkaa* "$PROJECT_HOME/$BUILD_DIR/" &&
     cd $BUILD_DIR &&
-    cmake -DAPP_NAME=$APP_NAME $KAA_TOOLCHAIN_PATH_SDK ..
-    make
+    cmake -DCMAKE_BUILD_TYPE=Debug \
```
Comment (rasendubi) [inline datacollectiondemo/source/cpp/build.sh]: Debug?

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp [hunk 3]
```diff
@@ -14,77 +14,110 @@
...
 
-    /*
-     * Stop the Kaa endpoint.
-     */
-    kaaClient->stop();
-
-    std::cout << "Data collection demo stopped" << std::endl;
+int main()
+{
+    TemperatureSensor sensor;
```
Comment (rasendubi) [inline datacollectiondemo/source/cpp/src/KaaDemo.cpp]: Scope can be reduced

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp [hunk 4]
```diff
@@ -14,77 +14,110 @@
...
-                                        + recordInfo.getRecordDeliveryTimeMs();
-
-            std::cout << "Received log record delivery info. Bucket Id [" <<  bucketInfo.getBucketId() << "]. "
-                      << "Record delivery time [" << timeSpent << " ms]." << std::endl;
-        } catch (std::exception& e) {
-            std::cout << "Exception was caught while waiting for callback future" << e.what();
-        }
+    void onConfigurationUpdated(const KaaRootConfiguration &configuration)
+    {
+        iostream_mutex_.lock();
```
Comment (rasendubi) [inline datacollectiondemo/source/cpp/src/KaaDemo.cpp]: why not guard?

Comment (ohamov) [reply datacollectiondemo/source/cpp/src/KaaDemo.cpp]: _Dead lock_
[operator <<](http://ru.cppreference.com/w/cpp/io/basic_ostream/operator_ltlt) lacks `noexcept` or `throw()` identifier. You are to use guard.

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp [hunk 5]
```diff
@@ -14,77 +14,110 @@
...
+private:
+    static constexpr auto savedConfig_ = "saved_config.cfg";
+    std::shared_ptr<IKaaClient> kaaClient_;
+    int32_t samplePeriod_;
+    boost::asio::io_service service_;
+    boost::posix_time::seconds interval_;
+    boost::asio::deadline_timer timer_;
+    std::mutex iostream_mutex_;
+
+    int32_t getTemperature(void)
```
Comment (ohamov) [inline datacollectiondemo/source/cpp/src/KaaDemo.cpp]: `void` can be omitted

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp [hunk 6]
```diff
@@ -14,77 +14,110 @@
...
-
-        futurePairs.push_back(std::make_pair(std::move(kaaClient->addLogRecord(logRecord)), logRecord.timeStamp));
-        std::cout << "Sent " << logNumber << "th record" << std::endl;
+        logRecord.temperature = getTemperature();
+        logRecord.timeStamp = std::time(nullptr);
+        // Send value of temperature
+        kaaClient_->addLogRecord(logRecord);
+        // Show log
+        iostream_mutex_.lock();
+        std::cout << "Sampled temperature: " << logRecord.temperature <<
```
Comment (GamovCoder) [inline datacollectiondemo/source/cpp/src/KaaDemo.cpp]: move trailing `<<` to the next line

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp [hunk 7]
```diff
@@ -14,77 +14,115 @@
...
-        futurePairs.push_back(std::make_pair(std::move(kaaClient->addLogRecord(logRecord)), logRecord.timeStamp));
-        std::cout << "Sent " << logNumber << "th record" << std::endl;
+        logRecord.temperature = getTemperature();
+        logRecord.timeStamp = std::time(nullptr);
+        // Send value of temperature
+        kaaClient_->addLogRecord(logRecord);
+        // Show log
+        {
+            std::lock_guard<std::mutex> guard(iostream_mutex_);
+            std::cout << "Sampled temperature: " << logRecord.temperature <<
```
Comment (ohamov) [inline datacollectiondemo/source/cpp/src/KaaDemo.cpp]: move trailing `<<` to the next line

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp [hunk 8]
```diff
@@ -14,77 +14,115 @@
...
-        } catch (std::exception& e) {
-            std::cout << "Exception was caught while waiting for callback future" << e.what();
+    void onConfigurationUpdated(const KaaRootConfiguration &configuration)
+    {
+        {
+            std::lock_guard<std::mutex> guard(iostream_mutex_);
+            std::cout << "Received configuration data. New sample period: "
+                << configuration.samplePeriod << " seconds" << "\n";
         }
+        samplePeriod_ = configuration.samplePeriod;
```
Comment (ohamov) [inline datacollectiondemo/source/cpp/src/KaaDemo.cpp:109]: move it before the `guard`. It allows you to drop unnecessary scope braces. Or just drop scope braces.

File: datacollectiondemo/source/cpp/build.sh (hunk 2 above)
Comment (RostakaGmfun) [reply datacollectiondemo/source/cpp/build.sh]: Looks like a tradition in Kaa sample applications :)

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp (hunk 1 above)
Comment (RostakaGmfun) [reply datacollectiondemo/source/cpp/src/KaaDemo.cpp:40]: Done

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp [hunk 9]
```diff
@@ -14,77 +14,111 @@
...
+        // Send value of temperature
+        kaaClient_->addLogRecord(logRecord);
+        // Show log
+        {
+            std::lock_guard<std::mutex> guard(iostream_mutex_);
+            std::cout << "Sampled temperature: " << logRecord.temperature
+                << ", timestamp: " << logRecord.timeStamp << std::endl;
         }
+        // Change timer expiry period
+        timer_.expires_from_now(boost::posix_time::seconds(samplePeriod_));
```
Comment (ohamov) [inline datacollectiondemo/source/cpp/src/KaaDemo.cpp]: In this case you have a delay, the period will not be quite accurate. You can avoid it this way:
`timer_->expires_at(timer_->expires_at() + boost::posix_time::seconds(samplePeriod_));`

File: datacollectiondemo/source/cpp/build.sh (hunk 2 above)
Comment (rasendubi) [reply datacollectiondemo/source/cpp/build.sh]: Would you like to break it?

File: datacollectiondemo/source/cpp/src/KaaDemo.cpp [hunk 10]
```diff
@@ -14,77 +14,111 @@
...
-        KaaUserLogRecord logRecord;
-        logRecord.level = kaa_log::Level::KAA_INFO;
-        logRecord.tag = "TAG";
-        logRecord.message = "MESSAGE_" + std::to_string(logNumber);
-        logRecord.timeStamp = TimeUtils::getCurrentTimeInMs();
+class TemperatureSensor : public IConfigurationReceiver {
+public:
+    TemperatureSensor():
+        kaaClient_(Kaa::newClient()),
+        interval_(samplePeriod_ = 0),
```
Comment (ohamov) [inline datacollectiondemo/source/cpp/src/KaaDemo.cpp]: one member initialization per line please

File: datacollectiondemo/source/cpp/build.sh [hunk 11]
```diff
@@ -58,50 +73,19 @@ function select_arch {
...
+          -DKAA_WITHOUT_EVENTS=1 \
+          -DKAA_WITHOUT_NOTIFICATIONS=1 \
+          -DKAA_WITHOUT_OPERATION_LONG_POLL_CHANNEL=1 \
+          -DKAA_WITHOUT_OPERATION_HTTP_CHANNEL=1 \
+          -DKAA_MAX_LOG_LEVEL=3 \
+           $KAA_TOOLCHAIN_PATH_SDK \
```
Comment (rasendubi) [inline datacollectiondemo/source/cpp/build.sh]: `"..."`

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by rasendubi** [comment]
@kaa-leeroy retest this please

**Comment by vadimol** [comment]
@kaa-leeroy retest this please

**Comment by vadimol** [comment]
There were issues on Jenkins. Now they are resolved.

**Comment by Oldnickus** [comment]
@kaa-leeroy test this please

**Comment by Oldnickus** [comment]
@kaa-leeroy test this please or die

**Pull request merged**